libraries/AP_HAL/DSP.cpp

```diff
@@ -227,4 +211,122 @@ float DSP::tau(const float x) const
+    uint16_t numpeaks = 0;
+    uint16_t peakX = 0;
+    float peakY = 0;
+    uint16_t pindex;
+    uint16_t xx[peakgroup];
+    float yy[peakgroup];
+    memset(xx, 0, peakgroup * sizeof(uint16_t));
+    memset(yy, 0, peakgroup * sizeof(float));
+
+    for (uint16_t j = (halfw << 1) - 2; j < length - smoothwidth - 1; j++) {
+        if (d[j] >= 0 && d[j + 1] <= 0 && !is_equal(d[j], d[j + 1])) { // detect zero crossing
+            if ((d[j] - d[j + 1]) > slopeThreshold) {
+                for (uint16_t k = 0; k < peakgroup; k++) {
+                    uint16_t groupIndex = j + k - n + 2;
+                    groupIndex = constrain_int16(groupIndex, 0, length - 1);
+                    xx[k] = groupIndex;
+                    yy[k] = input[groupIndex];
+                }
+                if (peakgroup < 3) {
+                    vector_max_float(yy, peakgroup, &peakY, &pindex);
+                } else {
+                    peakY = vector_mean_float(yy, peakgroup);
+                    pindex = val2index(yy, peakgroup, peakY);
+                }
+                peakX = xx[pindex];
+                //hal.console->printf("zero %d, gindex %d -> %d, index %d, val %f\n", j, j -n +2, j+peakgroup -1  - n +2, peakX, peakY);
+                // see if we have a valid peak
+                if (isfinite(peakY) && peakY >= ampThreshold) {
+                    // record in amplitude order
+                    for (int16_t i = 0; i < peaklen; i++) {
+                        if (i >= numpeaks) {
+                            peaks[i] = peakX;
+                            break;
+                        }
+                        if (peakY > input[peaks[i]]) {
+                            for (int16_t a = peaklen - 1; a > i; a--) {
+                                peaks[a] = peaks[a - 1];
+                            }
+                            peaks[i] = peakX;
+                            break;
+                        }
+                    }
+                    numpeaks++;
+                }
+            }
+        }
+    }
+
+    return numpeaks;
+}
+
+// Returns the index and the value of the element of a vector that is closest to val
+uint16_t DSP::val2index(const float* vector, uint16_t n, float val) const
+{
+    float minval = FLT_MAX;
+    uint16_t minidx = 0;
+    for (uint16_t i = 0; i < n; i++) {
+        float dif = fabsf(vector[i] - val);
+        if (dif < minval) {
+            minval = dif;
+            minidx = i;
+        }
+    }
+    return minidx;
+}
+
+// First derivative of vector using 2-point central difference.
+void DSP::derivative(const float* input, float* output, uint16_t n) const
+{
+    output[0] = input[1] - input[0];
+    output[n - 1] = input[n - 1] - input[n - 2];
+    for (uint16_t i = 1; i < n - 1; i++) {
+        output[i] = (input[i + 1] - input[i - 1]) / 2.0f;
+    }
+}
+
+// smooth a vector in-place
+void DSP::fastsmooth(float* input, uint16_t n, uint16_t smoothwidth) const
+{
+    float window[smoothwidth];
```

Comment:
use {} notation

Reply:
Can't do that here - the version of gcc we use (or clang maybe) in CI does not allow this for variable sized structures.

Reply:
interesting!